### PR TITLE
Fix:Static pod status is always Init:0/1 when init container GC'd before kubelet restart

### LINKED ISSUE #108537: Static pod status is always Init:0/1 when init container GC'd before kubelet restart.
### What happened?

When init container of static pod was GC'd and kubelet was restarted. The status of the static pod is alway Init:0/1 and never come back to Running.

```shell

kube-controller-manager-paas-192-168-16-112     0/1     Init:0/1                0          11d     192.168.16.112   paas-192-168-16-112   <none>           <none>
kube-controller-manager-paas-192-168-16-113     1/1     Running                 1          11d     192.168.16.113   paas-192-168-16-113   <none>           <none>

```

### What did you expect to happen?

The pod status is Running

### How can we reproduce it (as minimally and precisely as possible)?

1. remove exited init contianer of one static pod.
2. restart kubelet

### Anything else we need to know?

I noticed #96572 makes containers not restarting after init container GC'd. Non-static pod has no problem, but static seems to be affected by it.

### Kubernetes version

<details>

v1.22.1

</details>

### Cloud provider

<details>
Local Cluster
</details>

### OS version

<details>
</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by 249043822**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Since static pod has no podstatus saved except in mirror pod, so while init container GC'd  and kubelet restarts, static pod can only generate its pod status from current runtime, but after https://github.com/kubernetes/kubernetes/pull/96572 merged, init containers would not be re-created again, so we should reset init-container status for static pod correctly.
#### Which issue(s) this PR fixes:

Fixes #108537

#### Special notes for your reviewer:
/cc @sjenning @smarterclayton 
#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/kubelet/kubelet_pods.go [hunk 1]
```diff
@@ -1808,6 +1808,22 @@ func (kl *Kubelet) convertToAPIContainerStatuses(pod *v1.Pod, podStatus *kubecon
 		containerSeen[cName] = containerSeen[cName] + 1
 	}
 
+	// If any of the main containers have status and are Running, then all init containers must
+	// have been executed at some point in the past.  However, they could have been removed
+	// from the container runtime now, and if this is the case, we should set the state of init containers
+	// exited normally.
+	hasInitailized := false
```
Comment (VirrageS) [inline pkg/kubelet/kubelet_pods.go:1815]:
```suggestion
	hasInitialized := false
```

**Review by VirrageS — CHANGES_REQUESTED** [review]

**Review by yangjunmyfm192085 — COMMENTED** [review]
/priority important-soon
/triage accepted

**Comment by yangjunmyfm192085** [comment]
I think this pr solves a bug that has always existed

File: pkg/kubelet/kubelet_pods.go [hunk 2]
```diff
@@ -1808,6 +1808,22 @@ func (kl *Kubelet) convertToAPIContainerStatuses(pod *v1.Pod, podStatus *kubecon
...
+	// If any of the main containers have status and are Running, then all init containers must
+	// have been executed at some point in the past.  However, they could have been removed
+	// from the container runtime now, and if this is the case, we should set the state of init containers
+	// exited normally.
+	hasInitialized := false
+	if isInitContainer {
+		for i := range pod.Spec.Containers {
+			container := &pod.Spec.Containers[i]
+			status := podStatus.FindContainerStatusByName(container.Name)
+			if status != nil && status.State == kubecontainer.ContainerStateRunning {
```
Comment (matthyx) [inline pkg/kubelet/kubelet_pods.go:1820]: I am not convinced this is the right way to check, and probably we could get or compute this information elsewhere.

**Review by matthyx — CHANGES_REQUESTED** [review]
I think this reconciliation makes sense, given the new behavior introduced by https://github.com/kubernetes/kubernetes/pull/96572. I am not convinced by the `hasInitialized` detection.

File: pkg/kubelet/kubelet_pods.go (hunk 2 above)
Comment (249043822) [reply pkg/kubelet/kubelet_pods.go:1820]: Thanks @matthyx , there are already similar usages in other modules, https://github.com/kubernetes/kubernetes/blob/f33ca2306548719e5116b53fccfc278bffb809a8/pkg/kubelet/kuberuntime/kuberuntime_container.go#L842 and https://github.com/kubernetes/kubernetes/blob/f33ca2306548719e5116b53fccfc278bffb809a8/pkg/kubelet/status/status_manager.go#L386, mybe this is the only method to caculate this

Comment (matthyx) [reply pkg/kubelet/kubelet_pods.go:1820]: Ok, thanks... maybe this should be DRYed and moved as a function in `pod/utils.go`?

Comment (249043822) [reply pkg/kubelet/kubelet_pods.go:1820]: Sorry, I have not cared this before, I will rework it as what you commented

Comment (249043822) [reply pkg/kubelet/kubelet_pods.go:1820]: I think we can keep this change as it is now, because it is not completely the same with others

**Pull request closed without merging**